vsintegration/src/FSharp.Editor/Telemetry/TelemetryReporter.fs

```diff
@@ -3,25 +3,136 @@
 namespace Microsoft.VisualStudio.FSharp.Editor.Telemetry
 
 open Microsoft.VisualStudio.Telemetry
+open System
+open System.Diagnostics
+open System.Collections.Concurrent
+open Microsoft.VisualStudio.Shell
+open Microsoft.VisualStudio
+open Microsoft.VisualStudio.LanguageServices
+open Microsoft.VisualStudio.FSharp.Editor
 
+#nowarn "3220" // Ignore warning about direct tuple items access.
+
+[<RequireQualifiedAccess>]
+module TelemetryEvents =
+    [<Literal>]
+    let CodefixActivated = "codefixactivated"
+
+    [<Literal>]
+    let Hints = "hints"
+
+    [<Literal>]
+    let LanguageServiceStarted = "languageservicestarted"
+
+    [<Literal>]
+    let GetSymbolUsesInProjectsStarted = "getSymbolUsesInProjectsStarted"
+
+    [<Literal>]
+    let GetSymbolUsesInProjectsFinished = "getSymbolUsesInProjectsFinished"
+
+    [<Literal>]
+    let AddSyntacticCalssifications = "addsyntacticclassifications"
+
+    [<Literal>]
+    let AddSemanticCalssifications = "addsemanticclassifications"
+
+    [<Literal>]
+    let GetDiagnosticsForDocument = "getdiagnosticsfordocument"
+
+    [<Literal>]
+    let ProvideCompletions = "providecompletions"
+
+// TODO: needs to be something more sophisticated in future
+[<Struct; RequireQualifiedAccess; NoComparison; NoEquality>]
+type TelemetryThrottlingStrategy =
+    | NoThrottling
+    | Throttle of {| Timeout: TimeSpan |}
+    // At most, send one event per 3 seconds.
+    static member Default =
+        Throttle
+            {|
+                Timeout = TimeSpan.FromSeconds(3.0)
+            |}
+
+[<RequireQualifiedAccess>]
 module TelemetryReporter =
+    let internal noopDisposable =
+        { new IDisposable with
+            member _.Dispose() = ()
+        }
+
+    let internal lastSentEvents = ConcurrentDictionary<string, DateTime>()
```

Comment:
Maybe this could be called `lastTriggeredEvents` or `lastStartedEvents` instead. Since we track when the Event is initiated rather than when it's actually sent.